webhookValidator.go

```
@@ -0,0 +1,189 @@
+	errURLIsNotHTTPS         = errors.New("URL is not HTTPS")
+	errInvalidHost           = errors.New("Invalid host")
+	errIPGivenAsHost         = errors.New("Cannot use IP as host")
+	errLocalHostGivenAsHost  = errors.New("Cannot use Localhost as host")
+	errLoopbackGivenAsHost   = errors.New("Cannot use loopback host")
+)
+
+type Validator interface {
+	Validate(w Webhook) error
+}
+
+type Validators []Validator
+
+type ValidFunc func(Webhook) error
+
+type ValidURLFunc func(*url.URL) error
+
+// Validate runs the given webhook through each validator in the validators list.
+// It returns as soon as the webhook is considered invalid and returns nil if the
+// webhook is valid.
+func (vs Validators) Validate(w Webhook) error {
+	for _, v := range vs {
+		err := v.Validate(w)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GoodURL parses the given webhook's Config.URL, FailureURL, and Config.AlternativeURLs
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodURL(v []ValidURLFunc) ValidFunc {
+	return func(w Webhook) error {
+		//check Config.URL
+		parsedURL, err := url.ParseRequestURI(w.Config.URL)
+		if err != nil {
+			return fmt.Errorf("%w: %v", errInvalidURL, err)
+		}
+		for _, f := range v {
+			if f == nil {
+				continue
+			}
+			err = f(parsedURL)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidURL, err)
+			}
+		}
+		//check FailureURL if it exists
+		if w.FailureURL != "" {
+			parsedFailureURL, err := url.ParseRequestURI(w.FailureURL)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
+			}
+			for _, f := range v {
+				if f == nil {
+					continue
+				}
+				err = f(parsedFailureURL)
+				if err != nil {
+					return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
+				}
+			}
+		}
+		//check AlternativeURLs if they exist
+		if len(w.Config.AlternativeURLs) > 0 {
+			for _, u := range w.Config.AlternativeURLs {
+				if u != "" {
```

If we're given a slice, we expect the values in it to be non-empty.  If we find an empty string, we want to return an error rather than continuing.